Python/Python.sublime-syntax

```diff
@@ -454,39 +455,29 @@ contexts:
       captures:
         1: punctuation.definition.numeric.binary.python
     # complex
-    - match: (?i){{digitpart}}?(\.){{digitpart}}(?:e[\-\+]?{{digitpart}})?(j) # mandatory fraction
-      scope: constant.numeric.complex.python
-      captures:
-        1: punctuation.separator.decimal.python
-        2: storage.type.numeric.complex.python
-    - match: \b(?i)(?:{{digitpart}}(?:(\.){{digitpart}}?)?|(\.){{digitpart}})(?:e[\-\+]?{{digitpart}})(j) # mandatory exponent
-      scope: constant.numeric.complex.python
+    - match: |-
+        (?x:
+                      (\.)  {{digit}}  {{exponent}}? |
+          \b{{digit}} (\.)? {{digit}}? {{exponent}}?
+        )([Jj])
+      scope: constant.numeric.complex.imaginary.python
       captures:
         1: punctuation.separator.decimal.python
         2: punctuation.separator.decimal.python
-        3: storage.type.numeric.complex.python
-    - match: (?i){{digitpart}}(\.)?(j) # mandatory digitpart
-      scope: constant.numeric.complex.python
-      captures:
-        1: punctuation.separator.decimal.python
-        2: storage.type.numeric.complex.python
+        3: storage.type.numeric.imaginary.python
     # floating point
-    - match: (?i){{digitpart}}?(\.){{digitpart}}(?:e[\-\+]?{{digitpart}})? # mandatory fraction
-      scope: constant.numeric.float.python
-      captures:
-        1: punctuation.separator.decimal.python
-    - match: \b(?i)(?:{{digitpart}}(?:(\.){{digitpart}}?)?|(\.){{digitpart}})(?:e[\-\+]?{{digitpart}}) # mandatory exponent
-      scope: constant.numeric.float.python
+    - match: |-
+        (?x:
+                          (\.) {{digit}}  {{exponent}}? |
+          \b{{digit}} (?: (\.) {{digit}}? {{exponent}}? | {{exponent}} )
```

Comment:
The `|` being on the same column as above is kinda irritating because it's actually within a nested group. I'd put `\b{{digit}} {{exponent}}` on a third line.

Reply:
The idea is to prevent the parser from parsing a token twice. Therefore a dedicated `\b{{digit}} {{exponent}}` next to `\b{{digit}} (\.) {{digit}}? {{exponent}}?` would reduce the performance (slightly, but meassurable).


How about just ...

```yml
    - match: |-
        (?x:
          \b{{digits}} (?: (\.) {{digits}}? {{exponent}}? | {{exponent}} )
          | (\.) {{digits}} {{exponent}}?
        )
      scope: constant.numeric.float.decimal.python
      captures:
        1: punctuation.separator.decimal.python
        2: punctuation.separator.decimal.python
```